asf_heat_pump_suitability/research/exploratory/build_classification_model_block_of_flats.py

```diff
@@ -0,0 +1,798 @@
+all_uprns_df = all_uprns_df.with_columns(
+    n_stacked_uprns=pl.col("UPRN").count().over(["X_COORDINATE", "Y_COORDINATE"])
+)
+
+# Group by building and get the average and STD of UPRNs sharing the same coordinates
+agg_uprns_df = all_uprns_df.group_by("oct25_building_id").agg(
+    pl.col("apr25_building_id").first().name.keep(),
+    pl.col("n_stacked_uprns").mean().alias("avg_n_stacked_uprns"),
+    pl.col("n_stacked_uprns").std().alias("std_n_stacked_uprns"),
+)
+
+# Join all the calculated features together
+features_df = agg_building_df.join(
+    agg_uprns_df.select(
+        ["oct25_building_id", "avg_n_stacked_uprns", "std_n_stacked_uprns"]
+    ),
+    how="left",
+    on="oct25_building_id",
+)
+
+# %% [markdown]
+# ## Load and process labelled training data
+
+# %%
+# Group the manually labelled archetypes into block of flats or not
+block_of_flats_labels = ["BF", "BP", "SF"]
+non_block_labels = ["WB", "TE", "TF", "CO", "OF", "TT"]
+
+# %%
+labelled_datasets = dict()
+
+labelled_datasets["apr25"] = {
+    "reindeer_1": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_plymouth_buildings_containing_flats_sample_n200_seed10.kml",
+    "reindeer_2": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_small_plymouth_buildings_containing_flats_sample_n100_seed10.kml",
+}
+
+labelled_datasets["oct25"] = {
+    "reindeer_3": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_reindeer_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+    "raccoon": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_raccoon_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+    "anteater": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_anteater_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+    "leopard": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_leopard_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+    "springbok": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_springbok_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+    "yak": "s3://asf-heat-pump-suitability/local_heat_planning/labelling/labelled/CORRECTED_20251121_yak_LABELLED_buildings_containing_flats_sample_n147_seed10.kml",
+}
+
+
+# %%
+def extract_gdf_labelled_data(gdf, id_str):
+    """
+    Extract label, confidence, URL, and labeller from manually labelled sample data
+
+    Args:
+        gdf (gpd.GeoDataFrame): manually labelled sample data
+        id_str (str): name of building ID substring to search for in 'description' column of gdf
+
+    Returns:
+        gpd.GeoDataFrame: extracted information for manually labelled sample data
+    """
+    gdf[id_str] = gdf.description.str.extract(r"building_id: (.+) labeller")
```

Comment:
I'm getting an error when running this function below
`AttributeError: 'GeoDataFrame' object has no attribute 'description'`
Fix:
```suggestion
    gdf[id_str] = gdf.Description.str.extract(r"building_id: (.+) labeller")
```

Reply:
Sorry, this is because I upgraded my python version in the `conda` environment to 3.12 whereas the `environment.yaml` specifies 3.10. I upgraded it in order to test out something in `sklearn` that required an upgraded python version. The different versions of python cause the dataframes to be read in slightly differently (i.e. different capitalisation of column names). That's why you are getting this error but I didn't.